### PR TITLE
Avoid CharsRef Allocations in StreamInput

### PR DESCRIPTION
**Pull request by original-brownbear**
* Many messages deserialized from a `StreamInput` only contain short strings, some use-cases of instantiating a `StreamInput` don't deserialize any strings
  * Don't allocate `CharsRef` for small strings to save some allocations (especially on the IO threads)
  * Lazily allocate a larger `CharsRef` if needed for larger strings like we did before and have it live as long as the `StreamInput` like before as well

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -392,20 +392,35 @@ public Integer readOptionalVInt() throws IOException {
...
+    private static final int SMALL_STRING_LIMIT = 1024;
+
+    // Thread-local buffer for smaller strings
+    private static final ThreadLocal<CharsRef> smallSpare = ThreadLocal.withInitial(() -> new CharsRef(SMALL_STRING_LIMIT));
+
+    // Larger buffer used for long strings that can't fit into the thread-local buffer
+    // We don't use a CharsRefBuilder since we exactly know the size of the character array up front
     // this prevents calling grow for every character since we don't need this
-    private final CharsRef spare = new CharsRef();
+    private CharsRef largeSpare;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Can this be a SetOnce?

**Review by rjernst — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I'd rather not do that. Then we'd still needlessly create `SetOnce` (which seems about the same size of an empty `CharsRef`) here and even add a `volatile` read into the process.

**Comment by original-brownbear** [comment]
Thanks @rjernst !

**Pull request merged**